### PR TITLE
Small optimizations for the ExtensionManager

### PR DESCRIPTION
**Pull request by gvkries**
~~The initialization of the `ExtensionManager` is synchronous and must always be run before any method can be called. It can therefore be moved into the constructor to avoid a check before any method call. 
This is a minor performance optimization, but `EnsureInitialized` was called at least once during every request and on some requests dozens of times.~~

I based this optimization on PR #16379. 

Note: After discussion, only some minor optimizations are kept here.

### CODE REVIEW

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs [hunk 1]
```diff
@@ -16,382 +16,341 @@
...
-        }
+        L = logger;
 
-        public ILogger L { get; set; }
+        var modules = applicationContext.Application.Modules;
+        var loadedExtensions = new ConcurrentDictionary<string, ExtensionEntry>();
 
-        public IExtensionInfo GetExtension(string extensionId)
+        // Load all extensions in parallel
+        Parallel.ForEach(modules, (module, cancellationToken) =>
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:45]: Can we instead keep this logic in a method like `EnsureInitialized()` but call it in the constructor. You may want to called it `Initialize()`.

Also, converting the namespace scope to a file in a PR like this will make reviewing it very difficult. I suggest you revert the you change the namespace back so it's easy to review this PR. At some point in time, we'll convert all namespaces in the entire project. If this was a small class, I would not mind the namespace change

**Review by MikeAlhayek — COMMENTED** [review]
Have you done any performance benchmark to understand what is the performance gain here?

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs (hunk 1 above)
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:45]: Keeping the logic in a separate method would not allow to use `readonly` modifiers on the fields anymore. And I don't see how an `Initialize` method makes the code easier to follow.

**Comment by gvkries** [comment]
> Have you done any performance benchmark to understand what is the performance gain here?

I did not measure this, because I don't think it is worth the effort to test it. Same as e.g. making the classes `sealed`, which is also hardly measurable inside our code base, but safes some CPU cycles anways. And those tiny optimizations will add up over time.
Additionally, the motivation for this was more to make the code simpler.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs (hunk 1 above)
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs]: Wrapping the code in an Initialize method tells me that that code is being invoked is to initialize the class and will make it more readable and easier to follow. Placing so much code in the constructor makes the constructor very congested and hard to follow from maintenance perspective.

Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs]: I thought more about this and I'm going to change this back to an initialize method, but an explicit one. Doing so much work in the c'tor seems bad anyway.

Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs]: Hm, this is not going to be easy. Let me revert this change completely...

**Comment by gvkries** [comment]
I undid the initialization change and moved it back to the original `EnsureInitialize`. It's just a lot of work for a c'tor.

So there are only a few optimizations left here, e.g. sealing the class and removing some fields that don't need to be kept alive over the lifetime of the application.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs [hunk 2]
```diff
@@ -366,10 +372,10 @@ private bool HasDependency(IFeatureInfo f1, IFeatureInfo f2)
             return false;
         }
 
-        private int GetPriority(IFeatureInfo feature)
+        private static int GetPriority(IFeatureInfo feature, IExtensionPriorityStrategy[] extensionPriorityStrategies)
         {
             var sum = 0;
-            foreach (var strategy in _extensionPriorityStrategies)
+            foreach (var strategy in extensionPriorityStrategies)
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:378]: these 4 lines could be simplified with Linq.

```
var sum = extensionPriorityStrategies.Sum(strategy => strategy.GetPriority(feature))
```

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs [hunk 3]
```diff
@@ -152,7 +147,111 @@ public IEnumerable<IFeatureInfo> GetDependentFeatures(string featureId)
...
+                                // Type has no specific feature, add it to all features
+                                foreach (var curFeature in extension.ExtensionInfo.Features)
+                                {
+                                    typeFeatureProvider.TryAdd(exportedType, curFeature);
+                                }
+                            }
+                            else
+                            {
+                                typeFeatureProvider.TryAdd(exportedType, feature);
+                            }
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:222]: maybe drop the `else` here using something like this

```suggestion
                            if (feature != null)
                            {
                                typeFeatureProvider.TryAdd(exportedType, feature);
                            
                                continue;
                            }
                            
                            // Type has no specific feature, add it to all features.
                            foreach (var curFeature in extension.ExtensionInfo.Features)
                            {
                                typeFeatureProvider.TryAdd(exportedType, curFeature);
                            }
```

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs [hunk 4]
```diff
@@ -152,7 +147,111 @@ public IEnumerable<IFeatureInfo> GetDependentFeatures(string featureId)
...
+
+                    loadedExtensions.TryAdd(module.Name, entry);
+                });
+
+                // Get all types from all extension and add them to the type feature provider.
+                foreach (var loadedExtension in loadedExtensions)
+                {
+                    var extension = loadedExtension.Value;
+
+                    foreach (var exportedType in extension.ExportedTypes.Where(IsComponentType))
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:203]: Better to save the components once so it's not evaluated on each iteration. 
```suggestion
                var componentTypes = extension.ExportedTypes.Where(IsComponentType).ToArray();
                
                // Get all types from all extension and add them to the type feature provider.
                foreach (var componentType in componentTypes)
                {
                    var extension = componentType.Value;

                    foreach (var exportedType in componentTypes)
```

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs [hunk 5]
```diff
@@ -152,7 +147,111 @@ public IEnumerable<IFeatureInfo> GetDependentFeatures(string featureId)
...
+                // Load all extensions in parallel
+                Parallel.ForEach(modules, (module, cancellationToken) =>
+                {
+                    if (!module.ModuleInfo.Exists)
+                    {
+                        return;
+                    }
+
+                    var manifestInfo = new ManifestInfo(module.ModuleInfo);
+                    var extensionInfo = new ExtensionInfo(module.SubPath, manifestInfo, (mi, ei) =>
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:183]: what does `mi` and `ei` stand for? maybe better to give them a readable name

**Review by MikeAlhayek — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs [hunk 6]
```diff
@@ -166,16 +265,15 @@ public IEnumerable<IFeatureInfo> GetDependentFeatures(string featureId)
...
                     {
-                        dependencyIds.Add(dependency.Id);
                         stack.Push(GetFeatureDependenciesFunc(dependency, features));
                     }
                 }
             }
 
             // Preserve the underlying order of feature infos.
-            foreach (var featureInfo in _featureInfos)
+            foreach (var featureInfo in features)
```
[the remark this answers is not in the document]
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:276]: Done.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs [hunk 7]
```diff
@@ -198,16 +296,15 @@ public IEnumerable<IFeatureInfo> GetDependentFeatures(string featureId)
...
                     {
-                        dependencyIds.Add(dependency.Id);
                         stack.Push(GetDependentFeaturesFunc(dependency, features));
                     }
                 }
             }
 
             // Preserve the underlying order of feature infos.
-            foreach (var featureInfo in _featureInfos)
+            foreach (var featureInfo in features)
```
[the remark this answers is not in the document]
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:307]: Done.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs [hunk 8]
```diff
@@ -259,103 +349,19 @@ private static List<IFeatureInfo> GetFeatureDependenciesFunc(IFeatureInfo curren
...
-        }
-
-        private IFeatureInfo[] Order(IEnumerable<IFeatureInfo> featuresToOrder)
+        private static IFeatureInfo[] Order(IEnumerable<IFeatureInfo> featuresToOrder, IExtensionDependencyStrategy[] extensionDependencyStrategies, IExtensionPriorityStrategy[] extensionPriorityStrategies)
         {
             return featuresToOrder
                 .OrderBy(x => x.Id)
-                .OrderByDependenciesAndPriorities(HasDependency, GetPriority)
+                .OrderByDependenciesAndPriorities(
+                    (f1, f2) => HasDependency(f1, f2, extensionDependencyStrategies),
```
[the remark this answers is not in the document]
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:357]: Done.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs [hunk 9]
```diff
@@ -259,103 +349,19 @@ private static List<IFeatureInfo> GetFeatureDependenciesFunc(IFeatureInfo curren
...
+                    feature => GetPriority(feature, extensionPriorityStrategies))
                 .ToArray();
         }
 
-        private bool HasDependency(IFeatureInfo f1, IFeatureInfo f2)
+        private static bool HasDependency(IFeatureInfo f1, IFeatureInfo f2, IExtensionDependencyStrategy[] extensionDependencyStrategies)
         {
-            foreach (var s in _extensionDependencyStrategies)
+            foreach (var s in extensionDependencyStrategies)
```
[the remark this answers is not in the document]
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:364]: Done.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs [hunk 10]
```diff
@@ -152,7 +147,111 @@ public IEnumerable<IFeatureInfo> GetDependentFeatures(string featureId)
...
+                        }
+                    }
+                }
+
+                // Feature infos and entries are ordered by priority and dependencies.
+                _featureInfos = Order(
+                    loadedExtensions.SelectMany(extension => extension.Value.ExtensionInfo.Features),
+                    extensionDependencyStrategies as IExtensionDependencyStrategy[] ?? extensionDependencyStrategies.ToArray(),
+                    extensionPriorityStrategies as IExtensionPriorityStrategy[] ?? extensionPriorityStrategies.ToArray());
+                _features = _featureInfos.ToFrozenDictionary(f => f.Id, f => f);
```
[the remark this answers is not in the document]
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:232]: Done.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs [hunk 11]
```diff
@@ -152,7 +147,111 @@ public IEnumerable<IFeatureInfo> GetDependentFeatures(string featureId)
...
+
+                // Extensions are also ordered according to the weight of their first features.
+                _extensionsInfos = _featureInfos
+                    .Where(f => f.Id == f.Extension.Features.First().Id)
+                    .Select(f => f.Extension)
+                    .ToList();
```
[the remark this answers is not in the document]
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:238]: Done.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs (hunk 2 above)
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:378]: This would bring us a lot of additional allocations for removing 3 lines. I'd rather keep it as-is.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs (hunk 3 above)
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:222]: For me this makes it harder to read.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs (hunk 4 above)
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:203]: This is not necessary as it's iterated only once.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs (hunk 5 above)
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs]: Done.

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs (hunk 4 above)
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:203]: Its a loop inside another loop, meaning it could iterate more than once. Otherwise, why the nested loop?

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs (hunk 3 above)
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs:222]: okay. It's a preference. For me when I see multiple `else`, it makes the code harder to read. But, I understand if you like using else :)

File: src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs (hunk 2 above)
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs]: I trust you. Why why would this add more allocations?

Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs]: Because the Lambda expression needs a closure for the `feature` variable. And as `GetPriority` gets called a lot, those would add up.

Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs]: E.g. https://sharplab.io/#gist:d051810459d9fe5ba06dbf556292622a

**Review by MikeAlhayek — APPROVED** [review]
